File: src/main/julia/day03.jl
include("point.jl")

in1, in2 = map(s->split(s, ","), readlines("input03.txt"))[1:2]

function next(p::Point, d::Char)
    d == 'U' ? north(p) :
    d == 'D' ? south(p) :
    d == 'L' ? west(p) :
    east(p)
end

function wire(input)::Array{Point}
    p = Point(0, 0)
    points = []
    for leg = input
        dir = leg[1]
        len = parse(Int, leg[2:end])
        for _ in 1:len
            p = next(p, dir)
            push!(points, p)
        end        
        
    end
    return points
end

# part 1

w1, w2 = wire(in1), wire(in2)

i = collect(intersect(Set(w1), Set(w2)))

min(manhattan.(i)...)


# part 2

idx(p, a) = findfirst(x->x == p, a)

runlen = map(x->idx(x, w1) + idx(x, w2), i)

min(runlen...)

File: src/main/python/day03.py


with open("input03.txt") as f:
    lines = f.readlines()
    w1 = lines[0].split(",")
    w2 = lines[1].split(",")

# === part 1


def wirepoints(w):

    nextpoint = {
        'R': lambda p: (p[0]+1, p[1]),
        'L': lambda p: (p[0]-1, p[1]),        
        'U': lambda p: (p[0], p[1]-1),
        'D': lambda p: (p[0], p[1]+1),
    }

    p = (0,0)
    res = [p]
    for t in w:
        dir = t[0]
        run = int(t[1:])
        for _ in range(run):
            p = nextpoint[dir](p)
            res.append(p)
    return res

def manh(p):
    return abs(p[0]) + abs(p[1])

wp1 = wirepoints(w1)
wp2 = wirepoints(w2)

i1 = list(set(wp1) & set(wp2))
print(sorted(map(manh, i1))[1])


# === part 2

def steps(p):
    return wp1.index(p) + wp2.index(p)

print(sorted(map(steps, i1))[1])    